Sources/Subprocess/IO/AsyncIO+Darwin.swift

```diff
@@ -0,0 +1,165 @@
+//
+//===----------------------------------------------------------------------===//
+
+/// Darwin AsyncIO implementation based on DispatchIO
+
+// MARK: - macOS (DispatchIO)
+#if canImport(Darwin)
+
+#if canImport(System)
+@preconcurrency import System
+#else
+@preconcurrency import SystemPackage
+#endif
+
+internal import Dispatch
+
+final class AsyncIO: Sendable {
+    static let shared: AsyncIO = AsyncIO()
+
+    private init() {}
+
+    internal func read(
+        from diskIO: borrowing IOChannel,
+        upTo maxLength: Int
+    ) async throws -> DispatchData? {
+        return try await self.read(
+            from: diskIO.channel,
+            upTo: maxLength,
+        )
+    }
+
+    internal func read(
+        from dispatchIO: DispatchIO,
+        upTo maxLength: Int
+    ) async throws -> DispatchData? {
+        return try await withCheckedThrowingContinuation { continuation in
+            var buffer: DispatchData = .empty
+            dispatchIO.read(
+                offset: 0,
+                length: maxLength,
+                queue: .global()
+            ) { done, data, error in
+                if error != 0 {
+                    continuation.resume(
+                        throwing: SubprocessError(
+                            code: .init(.failedToReadFromSubprocess),
+                            underlyingError: .init(rawValue: error)
+                        )
+                    )
+                    return
+                }
+                if let data = data {
+                    if buffer.isEmpty {
+                        buffer = data
+                    } else {
+                        buffer.append(data)
+                    }
+                }
+                if done {
+                    if !buffer.isEmpty {
+                        continuation.resume(returning: buffer)
+                    } else {
+                        continuation.resume(returning: nil)
+                    }
+                }
+            }
+        }
+    }
+
+    #if SubprocessSpan
+    internal func write(
+        _ span: borrowing RawSpan,
+        to diskIO: borrowing IOChannel
+    ) async throws -> Int {
+        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Int, any Error>) in
+            let dispatchData = span.withUnsafeBytes {
+                return DispatchData(
+                    bytesNoCopy: $0,
+                    deallocator: .custom(
+                        nil,
+                        {
+                            // noop
+                        }
+                    )
+                )
+            }
+            self.write(dispatchData, to: diskIO) { writtenLength, error in
+                if let error = error {
+                    continuation.resume(throwing: error)
+                } else {
+                    continuation.resume(returning: writtenLength)
+                }
+            }
+        }
```

Comment:
As @weissi says above, you _mustn't_ escape the pointer from the span out of the `withUnsafeBytes`. You should rearrange the code as below:

```suggestion
            span.withUnsafeBytes {
                let dispatchData = DispatchData(
                    bytesNoCopy: $0,
                    deallocator: .custom(
                        nil,
                        {
                            // noop
                        }
                    )
                )
                self.write(dispatchData, to: diskIO) { writtenLength, error in
                    if let error {
                        continuation.resume(throwing: error)
                    } else {
                        continuation.resume(returning: writtenLength)
                    }
                }
            }
        }
```